cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -253,6 +253,19 @@ public ApplicationMeta getApplication(ApplicationId appId) throws IOException {
       .orElse(null);
   }
 
+  /**
+   * Gets the actual {@link ApplicationMeta} of the given application.
+   *
+   * @param appId the application ID to get the metadata
+   * @return the {@link ApplicationMeta} of the given application id, or {@code null} if no such application was found.
+   * @throws IOException if failed to read from the underlying {@link StructuredTable}
+   */
+  @Nullable
+  public ApplicationMeta getActualApplication(ApplicationId appId) throws IOException {
```

Comment:
I feel this method should be named as getApplication.  the original one can be changed to getLatestApplication, or might be even better to refactor it later as getApplication(ApplicationReference appReference)

Reply:
agreed

Reply:
hmmm, the original one actually gets the application matching passed in ID, only with the exception that it maps `-SNAPSHOT` to latest... I'm out of idea how to name them

Reply:
I see, yeah this is tricky, I think Albert mentioned store shouldn't be the place to handle the "-SNAPSHOT" to latest version. This should be done from Handler. Basically call getApplication(Appid) if version in appId is actual version, else call getLatest(AppReference), something like that

Reply:
agree with jingtao, should change the `getApplication(ApplicationId)` in this class to actually get that application. There should be another method to get the latest app, which takes an ApplicationReference as an argument.

This is one of the reasons that special mapping from `-SNAPSHOT` to latest should only happen at the very beginning. Putting it anywhere else makes the logic very difficult to follow, and very easy to introduce bugs.

Reply:
cool, working on that, since we have `ApplicationReference` merged